src/System.Memory/src/System/Buffers/ReadOnlySequence.cs

```diff
@@ -199,76 +400,128 @@ public ReadOnlySequence<T> Slice(long start, SequencePosition end)
         /// <param name="length">The length of the slice</param>
         public ReadOnlySequence<T> Slice(SequencePosition start, long length)
         {
-            BoundsCheck(start, _sequenceEnd); // check start before length
+            // Check start before length
+            uint sliceStartIndex = (uint)GetIndex(start);
+            object sliceStartObject = start.GetObject();
+
+            uint startIndex = (uint)GetIndex(_sequenceStart);
+            object startObject = _sequenceStart.GetObject();
+
+            uint endIndex = (uint)GetIndex(_sequenceEnd);
+            object endObject = _sequenceEnd.GetObject();
+
+            // Single-Segment Sequence
+            if (startObject == endObject)
```

Comment:
Here is bug for last segment. It should be changed some thing like next `if (sliceStartObject == endObject)`

Reply:
Why is that a bug? This is the check for IsSingleSegment.

Reply:
The bug is for the last segment. Code can be fixed here because for the last segment we can work just as for single segment.

Reply:
I don't understand. How do we know there is one segment or multiple segments? What do you mean by last segment?

SequencePosition start will not always be in the "last segment".

Reply:
I means that Slice called for the last segment.

Reply:
Bug is when start position is the last segment. Just call this method with next params: buffer.Slice(buffer.End, 0); and remove bug.

Reply:
> How do we know there is one segment or multiple segments?

It does not matter if Slice called for the last segment. The result will be single segment so we can works as it is single segmented sequence.

Reply:
> The same can be said about first segment though. So, why not `if (sliceStartObject == startObject)` ?

My optimization is incorrect because start will be checked by incorrect way.